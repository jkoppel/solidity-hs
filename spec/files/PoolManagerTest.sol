// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.20;

import {Test} from "forge-std/Test.sol";
import {IHooks} from "../src/interfaces/IHooks.sol";
import {Hooks} from "../src/libraries/Hooks.sol";
import {IPoolManager} from "../src/interfaces/IPoolManager.sol";
import {IProtocolFees} from "../src/interfaces/IProtocolFees.sol";
import {IProtocolFeeController} from "../src/interfaces/IProtocolFeeController.sol";
import {PoolManager} from "../src/PoolManager.sol";
import {TickMath} from "../src/libraries/TickMath.sol";
import {Pool} from "../src/libraries/Pool.sol";
import {Deployers} from "./utils/Deployers.sol";
import {Currency, CurrencyLibrary} from "../src/types/Currency.sol";
import {MockHooks} from "../src/test/MockHooks.sol";
import {MockContract} from "../src/test/MockContract.sol";
import {EmptyTestHooks} from "../src/test/EmptyTestHooks.sol";
import {PoolKey} from "../src/types/PoolKey.sol";
import {PoolModifyLiquidityTest} from "../src/test/PoolModifyLiquidityTest.sol";
import {BalanceDelta, BalanceDeltaLibrary} from "../src/types/BalanceDelta.sol";
import {PoolSwapTest} from "../src/test/PoolSwapTest.sol";
import {TestInvalidERC20} from "../src/test/TestInvalidERC20.sol";
import {GasSnapshot} from "forge-gas-snapshot/GasSnapshot.sol";
import {PoolEmptyUnlockTest} from "../src/test/PoolEmptyUnlockTest.sol";
import {Action} from "../src/test/PoolNestedActionsTest.sol";
import {PoolId} from "../src/types/PoolId.sol";
import {LPFeeLibrary} from "../src/libraries/LPFeeLibrary.sol";
import {Position} from "../src/libraries/Position.sol";
import {Constants} from "./utils/Constants.sol";
import {SafeCast} from "../src/libraries/SafeCast.sol";
import {AmountHelpers} from "./utils/AmountHelpers.sol";
import {ProtocolFeeLibrary} from "../src/libraries/ProtocolFeeLibrary.sol";
import {IProtocolFees} from "../src/interfaces/IProtocolFees.sol";
import {StateLibrary} from "../src/libraries/StateLibrary.sol";
import {Actions} from "../src/test/ActionsRouter.sol";

contract PoolManagerTest is Test, Deployers, GasSnapshot {
    Action[] _actions;

    function test_unlock_cannotBeCalledTwiceByCaller() public {
        _actions = [Action.NESTED_SELF_UNLOCK];
        nestedActionRouter.unlock(abi.encode(_actions));
    }

    function test_unlock_cannotBeCalledTwiceByDifferentCallers() public {
        _actions = [Action.NESTED_EXECUTOR_UNLOCK];
        nestedActionRouter.unlock(abi.encode(_actions));
    }

    // function testExtsloadForPoolPrice() public {
    //     IPoolManager.key = IPoolManager.PoolKey({
    //         currency0: currency0,
    //         currency1: currency1,
    //         fee: 100,
    //         hooks: IHooks(address(0)),
    //         tickSpacing: 10
    //     });
    //     manager.initialize(key, SQRT_PRICE_1_1, ZERO_BYTES);

    //     PoolId poolId = key.toId();
    //     bytes32 slot0Bytes = manager.extsload(keccak256(abi.encode(poolId, POOL_SLOT)));
    //     snapLastCall("poolExtsloadSlot0");

    //     uint160 sqrtPriceX96Extsload;
    //     assembly {
    //         sqrtPriceX96Extsload := and(slot0Bytes, sub(shl(160, 1), 1))
    //     }
    //     (uint160 sqrtPriceX96Slot0,,,,,) = manager.getSlot0(poolId);

    //     // assert that extsload loads the correct storage slot which matches the true slot0
    //     assertEq(sqrtPriceX96Extsload, sqrtPriceX96Slot0);
    // }

    // function testExtsloadMultipleSlots() public {
    //     IPoolManager.key = IPoolManager.PoolKey({
    //         currency0: currency0,
    //         currency1: currency1,
    //         fee: 100,
    //         hooks: IHooks(address(0)),
    //         tickSpacing: 10
    //     });
    //     manager.initialize(key, SQRT_PRICE_1_1, ZERO_BYTES);

    //     // populate feeGrowthGlobalX128 struct w/ modify + swap
    //     modifyLiquidityRouter.modifyLiquidity(key, IPoolManager.ModifyLiquidityParams(-120, 120, 5 ether, 0));
    //     swapRouter.swap(
    //         key,
    //         IPoolManager.SwapParams(false, 1 ether, TickMath.MAX_SQRT_PRICE - 1),
    //         PoolSwapTest.TestSettings(true, true)
    //     );
    //     swapRouter.swap(
    //         key,
    //         IPoolManager.SwapParams(true, 5 ether, TickMath.MIN_SQRT_PRICE + 1),
    //         PoolSwapTest.TestSettings(true, true)
    //     );

    //     PoolId poolId = key.toId();
    //     bytes memory value = manager.extsload(bytes32(uint256(keccak256(abi.encode(poolId, POOL_SLOT))) + 1), 2);
    //     snapLastCall("poolExtsloadTickInfoStruct");

    //     uint256 feeGrowthGlobal0X128Extsload;
    //     uint256 feeGrowthGlobal1X128Extsload;
    //     assembly {
    //         feeGrowthGlobal0X128Extsload := and(mload(add(value, 0x20)), sub(shl(256, 1), 1))
    //         feeGrowthGlobal1X128Extsload := and(mload(add(value, 0x40)), sub(shl(256, 1), 1))
    //     }

    //     assertEq(feeGrowthGlobal0X128Extsload, 408361710565269213475534193967158);
    //     assertEq(feeGrowthGlobal1X128Extsload, 204793365386061595215803889394593);
    // }

    function test_getPosition() public view {
        (uint128 liquidity,,) = manager.getPositionInfo(key.toId(), address(modifyLiquidityRouter), -120, 120, 0);
        assert(LIQUIDITY_PARAMS.liquidityDelta > 0);
        assertEq(liquidity, uint128(uint256(LIQUIDITY_PARAMS.liquidityDelta)));
    }

    function supportsInterface(bytes4) external pure returns (bool) {
        return true;
    }
}
