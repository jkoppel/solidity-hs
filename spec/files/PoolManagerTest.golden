[ Pragma
    ( PragmaDefinition
        { pragma = "solidity ^0.8.20" }
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "Test"
                , Nothing
                )
            ] "forge-std/Test.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "IHooks"
                , Nothing
                )
            ] "../src/interfaces/IHooks.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "Hooks"
                , Nothing
                )
            ] "../src/libraries/Hooks.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "IPoolManager"
                , Nothing
                )
            ] "../src/interfaces/IPoolManager.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "IProtocolFees"
                , Nothing
                )
            ] "../src/interfaces/IProtocolFees.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "IProtocolFeeController"
                , Nothing
                )
            ] "../src/interfaces/IProtocolFeeController.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "PoolManager"
                , Nothing
                )
            ] "../src/PoolManager.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "TickMath"
                , Nothing
                )
            ] "../src/libraries/TickMath.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "Pool"
                , Nothing
                )
            ] "../src/libraries/Pool.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "Deployers"
                , Nothing
                )
            ] "./utils/Deployers.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "Currency"
                , Nothing
                )
            ,
                ( Identifier "CurrencyLibrary"
                , Nothing
                )
            ] "../src/types/Currency.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "MockHooks"
                , Nothing
                )
            ] "../src/test/MockHooks.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "MockContract"
                , Nothing
                )
            ] "../src/test/MockContract.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "EmptyTestHooks"
                , Nothing
                )
            ] "../src/test/EmptyTestHooks.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "PoolKey"
                , Nothing
                )
            ] "../src/types/PoolKey.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "PoolModifyLiquidityTest"
                , Nothing
                )
            ] "../src/test/PoolModifyLiquidityTest.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "BalanceDelta"
                , Nothing
                )
            ,
                ( Identifier "BalanceDeltaLibrary"
                , Nothing
                )
            ] "../src/types/BalanceDelta.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "PoolSwapTest"
                , Nothing
                )
            ] "../src/test/PoolSwapTest.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "TestInvalidERC20"
                , Nothing
                )
            ] "../src/test/TestInvalidERC20.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "GasSnapshot"
                , Nothing
                )
            ] "forge-gas-snapshot/GasSnapshot.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "PoolEmptyUnlockTest"
                , Nothing
                )
            ] "../src/test/PoolEmptyUnlockTest.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "Action"
                , Nothing
                )
            ] "../src/test/PoolNestedActionsTest.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "PoolId"
                , Nothing
                )
            ] "../src/types/PoolId.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "LPFeeLibrary"
                , Nothing
                )
            ] "../src/libraries/LPFeeLibrary.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "Position"
                , Nothing
                )
            ] "../src/libraries/Position.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "Constants"
                , Nothing
                )
            ] "./utils/Constants.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "SafeCast"
                , Nothing
                )
            ] "../src/libraries/SafeCast.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "AmountHelpers"
                , Nothing
                )
            ] "./utils/AmountHelpers.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "ProtocolFeeLibrary"
                , Nothing
                )
            ] "../src/libraries/ProtocolFeeLibrary.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "IProtocolFees"
                , Nothing
                )
            ] "../src/interfaces/IProtocolFees.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "StateLibrary"
                , Nothing
                )
            ] "../src/libraries/StateLibrary.sol"
        )
    )
, Import
    ( ImportDefinition
        ( ImportAliases
            [
                ( Identifier "Actions"
                , Nothing
                )
            ] "../src/test/ActionsRouter.sol"
        )
    )
, Contract
    ( ContractDefinition
        { abstract = False
        , name = Identifier "PoolManagerTest"
        , inheritance = Just
            [ InheritanceSpecifier
                ( IdentifierPath
                    [ Identifier "Test" ]
                ) Nothing
            , InheritanceSpecifier
                ( IdentifierPath
                    [ Identifier "Deployers" ]
                ) Nothing
            , InheritanceSpecifier
                ( IdentifierPath
                    [ Identifier "GasSnapshot" ]
                ) Nothing
            ]
        , body =
            [ CStateVariableDec
                ( StateVariableDec
                    { kind = ArrayType
                        ( IdentifierType
                            ( IdentifierPath
                                [ Identifier "Action" ]
                            )
                        ) [ ArrayTypeEmpty ]
                    , modifiers = Just []
                    , ident = Identifier "_actions"
                    , expr = Nothing
                    }
                )
            , CFunction
                ( FunctionDefinition
                    { name = Identifier "test_unlock_cannotBeCalledTwiceByCaller"
                    , params = []
                    , restrictions = [ FuncVisibility FuncPublic ]
                    , returns = Nothing
                    , body = BlockStatement
                        [ ExpressionStatement
                            ( BinaryExpression Assign
                                ( IdentifierExpression
                                    ( Identifier "_actions" )
                                )
                                ( InlineArrayExpression
                                    [ MemberAccess
                                        ( IdentifierExpression
                                            ( Identifier "Action" )
                                        )
                                        ( MemberAccessIdentifier
                                            ( Identifier "NESTED_SELF_UNLOCK" )
                                        )
                                    ]
                                )
                            )
                        , ExpressionStatement
                            ( FunctionCall
                                ( MemberAccess
                                    ( IdentifierExpression
                                        ( Identifier "nestedActionRouter" )
                                    )
                                    ( MemberAccessIdentifier
                                        ( Identifier "unlock" )
                                    )
                                )
                                ( CommaArguments
                                    [ FunctionCall
                                        ( MemberAccess
                                            ( IdentifierExpression
                                                ( Identifier "abi" )
                                            )
                                            ( MemberAccessIdentifier
                                                ( Identifier "encode" )
                                            )
                                        )
                                        ( CommaArguments
                                            [ IdentifierExpression
                                                ( Identifier "_actions" )
                                            ]
                                        )
                                    ]
                                )
                            )
                        ]
                    }
                )
            , CFunction
                ( FunctionDefinition
                    { name = Identifier "test_unlock_cannotBeCalledTwiceByDifferentCallers"
                    , params = []
                    , restrictions = [ FuncVisibility FuncPublic ]
                    , returns = Nothing
                    , body = BlockStatement
                        [ ExpressionStatement
                            ( BinaryExpression Assign
                                ( IdentifierExpression
                                    ( Identifier "_actions" )
                                )
                                ( InlineArrayExpression
                                    [ MemberAccess
                                        ( IdentifierExpression
                                            ( Identifier "Action" )
                                        )
                                        ( MemberAccessIdentifier
                                            ( Identifier "NESTED_EXECUTOR_UNLOCK" )
                                        )
                                    ]
                                )
                            )
                        , ExpressionStatement
                            ( FunctionCall
                                ( MemberAccess
                                    ( IdentifierExpression
                                        ( Identifier "nestedActionRouter" )
                                    )
                                    ( MemberAccessIdentifier
                                        ( Identifier "unlock" )
                                    )
                                )
                                ( CommaArguments
                                    [ FunctionCall
                                        ( MemberAccess
                                            ( IdentifierExpression
                                                ( Identifier "abi" )
                                            )
                                            ( MemberAccessIdentifier
                                                ( Identifier "encode" )
                                            )
                                        )
                                        ( CommaArguments
                                            [ IdentifierExpression
                                                ( Identifier "_actions" )
                                            ]
                                        )
                                    ]
                                )
                            )
                        ]
                    }
                )
            , CFunction
                ( FunctionDefinition
                    { name = Identifier "test_getPosition"
                    , params = []
                    , restrictions =
                        [ FuncVisibility FuncPublic
                        , FuncMutability View
                        ]
                    , returns = Nothing
                    , body = BlockStatement
                        [ VariableStatement
                            ( VariableDecTupleStatement
                                [ Just
                                    ( VariableDeclaration
                                        { kind = ElementaryType
                                            ( UnsignedInteger
                                                ( Just 128 )
                                            )
                                        , location = Nothing
                                        , ident = Identifier "liquidity"
                                        }
                                    )
                                , Nothing
                                , Nothing
                                ]
                                ( FunctionCall
                                    ( MemberAccess
                                        ( IdentifierExpression
                                            ( Identifier "manager" )
                                        )
                                        ( MemberAccessIdentifier
                                            ( Identifier "getPositionInfo" )
                                        )
                                    )
                                    ( CommaArguments
                                        [ FunctionCall
                                            ( MemberAccess
                                                ( IdentifierExpression
                                                    ( Identifier "key" )
                                                )
                                                ( MemberAccessIdentifier
                                                    ( Identifier "toId" )
                                                )
                                            )
                                            ( CommaArguments [] )
                                        , FunctionCall
                                            ( TypeExpression ( ElementaryType Address ) )
                                            ( CommaArguments
                                                [ IdentifierExpression
                                                    ( Identifier "modifyLiquidityRouter" )
                                                ]
                                            )
                                        , UnaryExpression UPreSub
                                            ( ExpressionLiteral
                                                ( NumberLiteral 120.0 Nothing )
                                            )
                                        , ExpressionLiteral
                                            ( NumberLiteral 120.0 Nothing )
                                        , ExpressionLiteral
                                            ( NumberLiteral 0.0 Nothing )
                                        ]
                                    )
                                )
                            )
                        , ExpressionStatement
                            ( FunctionCall
                                ( IdentifierExpression
                                    ( Identifier "assert" )
                                )
                                ( CommaArguments
                                    [ BinaryExpression GreaterThan
                                        ( MemberAccess
                                            ( IdentifierExpression
                                                ( Identifier "LIQUIDITY_PARAMS" )
                                            )
                                            ( MemberAccessIdentifier
                                                ( Identifier "liquidityDelta" )
                                            )
                                        )
                                        ( ExpressionLiteral
                                            ( NumberLiteral 0.0 Nothing )
                                        )
                                    ]
                                )
                            )
                        , ExpressionStatement
                            ( FunctionCall
                                ( IdentifierExpression
                                    ( Identifier "assertEq" )
                                )
                                ( CommaArguments
                                    [ IdentifierExpression
                                        ( Identifier "liquidity" )
                                    , FunctionCall
                                        ( TypeExpression
                                            ( ElementaryType
                                                ( UnsignedInteger
                                                    ( Just 128 )
                                                )
                                            )
                                        )
                                        ( CommaArguments
                                            [ FunctionCall
                                                ( TypeExpression
                                                    ( ElementaryType
                                                        ( UnsignedInteger
                                                            ( Just 256 )
                                                        )
                                                    )
                                                )
                                                ( CommaArguments
                                                    [ MemberAccess
                                                        ( IdentifierExpression
                                                            ( Identifier "LIQUIDITY_PARAMS" )
                                                        )
                                                        ( MemberAccessIdentifier
                                                            ( Identifier "liquidityDelta" )
                                                        )
                                                    ]
                                                )
                                            ]
                                        )
                                    ]
                                )
                            )
                        ]
                    }
                )
            , CFunction
                ( FunctionDefinition
                    { name = Identifier "supportsInterface"
                    , params =
                        [ Parameter
                            { kind = ElementaryType
                                ( FixedBytes 4 )
                            , location = Nothing
                            , ident = Nothing
                            }
                        ]
                    , restrictions =
                        [ FuncVisibility FuncExternal
                        , FuncMutability Pure
                        ]
                    , returns = Just
                        [ Parameter
                            { kind = ElementaryType Bool
                            , location = Nothing
                            , ident = Nothing
                            }
                        ]
                    , body = BlockStatement
                        [ Return
                            ( Just
                                ( ExpressionLiteral ( BooleanLiteral True ) )
                            )
                        ]
                    }
                )
            ]
        }
    )
]